src/lib.rs

```diff
@@ -98,16 +102,66 @@ impl Diagnostic {
         }
     }
 }
-
-pub type Sources<'a> = [&'a dyn SourceContainer];
+pub trait SourceLocation {
+    /// returns the location of this source-container. Used when reporting diagnostics.
+    fn get_location(&self) -> Option<&str>;
```

Comment:
as discussed, lets force a location.
at the moment we don't see a situation where no location-string would be better than a dummy location string (e.g. "<in_memory>" or "<unit_test>".)

Reply:
Traits are merged back, and we now have a method to get a String out of the reader.